File: mfish/src/main/java/com/mordrum/mfish/server/FishingLootGenerator.java
package com.mordrum.mfish.server;

import com.mordrum.mfish.common.CommonProxy;
import com.mordrum.mfish.common.Fish;
import com.mordrum.mfish.common.events.FishCaughtEvent;
import net.minecraft.entity.Entity;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.entity.player.EntityPlayerMP;
import net.minecraft.entity.projectile.EntityFishHook;
import net.minecraft.init.Blocks;
import net.minecraft.item.ItemStack;
import net.minecraft.nbt.NBTTagCompound;
import net.minecraft.stats.StatisticsManagerServer;
import net.minecraft.world.biome.Biome;
import net.minecraftforge.common.MinecraftForge;
import org.apache.commons.math3.distribution.FDistribution;

import java.math.BigDecimal;
import java.math.MathContext;
import java.text.DecimalFormat;
import java.util.Arrays;
import java.util.Random;
import java.util.Set;

public class FishingLootGenerator {
    private static FDistribution fDistribution = new FDistribution(1, 9.8);

    public static ItemStack getFishingLoot(EntityPlayer player, Entity fishHook) {
        Biome biome = fishHook.getEntityWorld().getBiome(fishHook.getPosition());

        Set<Fish> fishForBiome = Fish.getFishForBiome(biome);

		if (fishForBiome.isEmpty()) {
			return new ItemStack(Blocks.COBBLESTONE);
		}

        int rollMax = fishForBiome.stream().mapToInt(Fish::getRarity).sum();

        Random random = new Random();
        int roll = random.nextInt(rollMax);

        FDistribution fDistribution = new FDistribution(0.89, 100);
        int i = 0;
        for (Fish fish : fishForBiome) {
            i += fish.getRarity();
            if (roll <= i) {
                double originalWeight = fDistribution.sample() * 10;
                double weight = originalWeight;
                String prefix = null;
                if (originalWeight < 1) {
                    weight = 1 + Math.random();
                    prefix = "Tiny";
                } else if (originalWeight > 30) {
                    prefix = "Large";
                } else if (originalWeight > 75) {
                    prefix = "Huge";
                } else if (originalWeight > 150) {
                    prefix = "Massive";
                }
                weight = new BigDecimal(weight).round(new MathContext(3)).doubleValue();
                ItemStack itemStack = new ItemStack(CommonProxy.rawItem, 1, fish.getMetadata());
                if (!itemStack.hasTagCompound()) itemStack.setTagCompound(new NBTTagCompound());
                itemStack.getTagCompound().setDouble("Weight", weight);
                if (prefix != null) itemStack.getTagCompound().setString("Prefix", prefix);

                //TODO decide if this is the best place to call these
                MinecraftForge.EVENT_BUS.post(new FishCaughtEvent(player, (EntityFishHook) fishHook, fish, weight));
                HighscoreManager.checkHighscore(player, fish, weight);


                /*FIXME advancements
	            StatisticsManagerServer statFile = ((EntityPlayerMP) player).getStatFile();
	            if (!statFile.hasAchievementUnlocked(Achievements.FISHOLOGIST)) {
		            JsonSerializableSet jsonserializableset = statFile.getProgress(Achievements.FISHOLOGIST);
					if (jsonserializableset == null) {
						jsonserializableset = statFile.setProgress(Achievements.FISHOLOGIST, new JsonSerializableSet());
					}
					jsonserializableset.add(fish.getName());
					if (jsonserializableset.size() >= 30) {
						player.addStat(Achievements.FISHOLOGIST);
					} else if (jsonserializableset.size() >= 10) {
						player.addStat(Achievements.ANGLER);
					} else if (jsonserializableset.size() >= 3) {
						player.addStat(Achievements.FISHING_101);
					}
	            }
	            */

	            /*FIXME advancements
	            // Big catch achievement
	            if (weight >= 100.0) {
					player.addStat(Achievements.THE_BIG_ONE);
	            } else if (weight >= 30.0) {
	            	player.addStat(Achievements.BIG_CATCH);
	            } else if (originalWeight <= 1.0) {
		            player.addStat(Achievements.LITTLE_GUPPY);
	            }
	            */

	            // Keep track of the number of pounds this player has caught so far
	            NBTTagCompound entityData = player.getEntityData();
	            double totalWeightCaughtSoFar = 0.0;
	            if (entityData.hasKey("totalweight")) totalWeightCaughtSoFar = entityData.getDouble("totalweight");
				totalWeightCaughtSoFar += weight;
				entityData.setDouble("totalweight", totalWeightCaughtSoFar);

				/*FIXME advancements
                if (totalWeightCaughtSoFar >= 100) {
                    player.addStat(Achievements.TO_FEED_A_FAMILY, 1);
                } else if (totalWeightCaughtSoFar >= 1000) {
	                player.addStat(Achievements.TO_FEED_A_VILLAGE, 1);
                } else if (totalWeightCaughtSoFar >= 10000) {
	                player.addStat(Achievements.TO_FEED_AN_ARMY, 1);
                }
                */

	            return itemStack;
            }
        }

        return new ItemStack(Blocks.COBBLESTONE, 1);
    }

    public static void main(String[] args) {
        fDistribution = new FDistribution(0.89, 100);
        calculateForSample(fDistribution.sample(100000));
    }

    private static void calculateForSample(double[] sample) {
        System.out.println("NDGF: " + fDistribution.getNumeratorDegreesOfFreedom());

        double avg = 0;
        for (int i = 0; i < sample.length; i++) {
            double v = sample[i] * 10;
            if (v < 1) v = 1 + Math.random();
            if (v > 250) v = 250;
            sample[i] = v;

            avg += v;
        }

        Arrays.sort(sample);
        DecimalFormat df = new DecimalFormat("#.##");
        System.out.println("Random: " + df.format(sample[new Random().nextInt(sample.length)]));
        System.out.println("Smallest " + df.format(sample[0]));
        System.out.println("Largest " + df.format(sample[sample.length - 1]));
        System.out.println("Median: " + df.format(sample[sample.length / 2]));
        System.out.println("Mean: " + df.format((avg / sample.length)));
    }
}

File: mfish/src/main/java/com/mordrum/mfish/server/FishingLootGenerator.kt
package com.mordrum.mfish.server

import com.mordrum.mfish.common.CommonProxy
import com.mordrum.mfish.common.Fish
import com.mordrum.mfish.common.events.FishCaughtEvent
import net.minecraft.entity.Entity
import net.minecraft.entity.player.EntityPlayer
import net.minecraft.entity.projectile.EntityFishHook
import net.minecraft.init.Blocks
import net.minecraft.item.ItemStack
import net.minecraft.nbt.NBTTagCompound
import net.minecraft.util.text.TextComponentString
import net.minecraftforge.common.MinecraftForge
import org.apache.commons.math3.distribution.FDistribution
import java.math.BigDecimal
import java.math.MathContext
import java.text.DecimalFormat
import java.util.*
import java.util.function.Consumer

object FishingLootGenerator {
    private val sb = StringBuilder()
    private val decimalFormat = DecimalFormat("#,###.##")

    fun getFishingLoot(player: EntityPlayer, fishHook: Entity): ItemStack {
        // Get the fish available in the biome where the player is fishing
        val fishForBiome = Fish.getFishForBiome(fishHook.entityWorld.getBiome(fishHook.position))

        // Add up the rarity of all the fish, this will tell us the maximum our random number can be
        val rollMax = fishForBiome.map { it.rarity }.sum()
        // Create a random number generator, then roll a number to determine what fish will be selected
        val roll = Random().nextInt(rollMax)

        // Go through the list of fish and pull out the fish that was selected
        var i = 0
        var selectedFish: Fish? = null
        for (fish in fishForBiome) {
            i += fish.rarity
            if (roll <= i) {
                selectedFish = fish
                break
            }
        }
        // If no fish was selected (somehow) (usually the result of a bad config file), return cobblestone
        if (selectedFish == null) return ItemStack(Blocks.COBBLESTONE)

        // Get a random weight for the fish and determine its sized based prefix
        val weight: Double = rollFishWeight()
        val prefix: String
        if (weight > 150) prefix = "Massive"
        else if (weight > 75) prefix = "Huge"
        else if (weight > 30) prefix = "Large"
        else if (weight > 1) prefix = ""
        else prefix = "Baby"

        // Create an NBT tag compound so the fish itemstack will keep track of its data
        val tagCompound = NBTTagCompound()
        tagCompound.setDouble("weight", weight)
        tagCompound.setString("prefix", prefix)

        val itemStack = ItemStack(CommonProxy.rawItem, 1, selectedFish.metadata)
        itemStack.tagCompound = tagCompound

        // Post a FishCaughtEvent to the eventbus so things can interact with this method
        MinecraftForge.EVENT_BUS.post(FishCaughtEvent(player, fishHook as EntityFishHook, selectedFish, weight))

        //TODO migrate these methods out, perhaps into a listener class that listens for FishCaughtEvent
        APIHelper.checkHighscore(player, selectedFish, weight, Consumer { isHighscore ->
            if (isHighscore) {
                sb.setLength(0)
                sb.append(player.name)
                        .append(" just caught a record breaking ")
                        .append(decimalFormat.format(weight))
                        .append("lb ")
                        .append(selectedFish!!.name)
                        .append("!")
                val component = TextComponentString(sb.toString())
                for (entityPlayerMP in player.server!!.playerList.players) {
                    entityPlayerMP.sendMessage(component)
                }
            } else {
                player.sendMessage(TextComponentString("You caught a " + decimalFormat.format(weight) + "lb " + selectedFish!!.name))
            }
        }, Consumer { error ->
            player.sendMessage(TextComponentString("An error occurred while sending your fish to the API server: " + error.message))
        })
        //FIXME advancements
        //checkAchievements(player, selectedFish, weight)

        return itemStack
    }

    private fun rollFishWeight(): Double {
        val fDistribution = FDistribution(0.89, 100.0)
        var weight = fDistribution.sample() * 10

        // If the fish is really tiny, turn it into a baby fish that weighs between 0.5 and 1.0 lbs
        if (weight < 1) weight = Math.random() / 2 + 0.5
        // Trim off excess decimal places
        return BigDecimal(weight).round(MathContext(3)).toDouble()
    }

    /*FIXME advancements
    private fun checkAchievements(player: EntityPlayer, fish: Fish, weight: Double) {
        val statFile = (player as EntityPlayerMP).statFile
        if (!statFile.hasAchievementUnlocked(Achievements.FISHOLOGIST)) {
            var jsonserializableset: JsonSerializableSet? = statFile.getProgress<JsonSerializableSet>(Achievements.FISHOLOGIST)
            if (jsonserializableset == null) {
                jsonserializableset = statFile.setProgress(Achievements.FISHOLOGIST, JsonSerializableSet())
            }
            jsonserializableset!!.add(fish.name)
            if (jsonserializableset.size >= 30) {
                player.addStat(Achievements.FISHOLOGIST)
            } else if (jsonserializableset.size >= 10) {
                player.addStat(Achievements.ANGLER)
            } else if (jsonserializableset.size >= 3) {
                player.addStat(Achievements.FISHING_101)
            }
        }

        // Big catch achievement
        if (weight >= 100.0) {
            player.addStat(Achievements.THE_BIG_ONE)
        } else if (weight >= 30.0) {
            player.addStat(Achievements.BIG_CATCH)
        } else if (weight <= 1.0) {
            player.addStat(Achievements.LITTLE_GUPPY)
        }

        // Keep track of the number of pounds this player has caught so far
        val entityData = player.getEntityData()
        var totalWeightCaughtSoFar = 0.0
        if (entityData.hasKey("totalweight")) totalWeightCaughtSoFar = entityData.getDouble("totalweight")
        totalWeightCaughtSoFar += weight
        entityData.setDouble("totalweight", totalWeightCaughtSoFar)
        if (totalWeightCaughtSoFar >= 100) {
            player.addStat(Achievements.TO_FEED_A_FAMILY, 1)
        } else if (totalWeightCaughtSoFar >= 1000) {
            player.addStat(Achievements.TO_FEED_A_VILLAGE, 1)
        } else if (totalWeightCaughtSoFar >= 10000) {
            player.addStat(Achievements.TO_FEED_AN_ARMY, 1)
        }
    }
    */
}
